### PR TITLE
Adding Multimodal Abilities to Crew

### PR DESCRIPTION
**Pull request by joaomdmoura**
This PR:
- Adds a series of new test to cover overall tools features
- Fix bugs around tools being overwritten by combination of factors (agent tools vs task tools vs code execution vs allow delegation)
- Adds the ability to multimodal agents

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment for CrewAI PR #1805 

## Overview
This pull request introduces significant enhancements to the CrewAI framework, specifically focusing on the handling of delegation tools, multimodal capabilities, and general code organization improvements. This review highlights key changes, suggests actionable improvements, and emphasizes historical context and implications.

## Key Changes
1. **Agent Class Enhancements**
   - Addition of `multimodal` flag in the Agent class.
   - New support for multimodal tools through the method `get_multimodal_tools()`.
   - Improved handling for code execution tools.

2. **Core Changes to CrewAI**
   - Refactored logic for tool preparation and delegation.
   - Better support for multimodal agents and tools.
   - Enhanced inheritance and override behavior for tools.

## Detailed Analysis & Recommendations

### 1. Tool Management Refactoring

**Current Implementation Example:**
```python
def _prepare_tools(self, agent: BaseAgent, task: Task, tools: List[Tool]):
```

**Recommended Improvement:**
- Introduce a structured approach for preparing tools, enhancing maintainability.

```python
def _prepare_tools(self, agent: BaseAgent, task: Task, tools: List[Tool]) -> List[Tool]:
    ...
```
- **Documentation**: Add docstrings to clarify arguments and return types, improving readability.

### 2. Multimodal Support Implementation

**Current Implementation Example:**
```python
def get_multimodal_tools(self):
```

**Recommended Improvement:**
- Ensure that the method returns a list of multimodal tools clearly indicated with type hints.

```python
def get_multimodal_tools(self) -> List[Tool]:
    ...
```
- Consider expanding this for future multimodal tools and providing documentation.

### 3. Tool Response Handling

**Current Implementation Example:**
```python
if formatted_answer.tool == "Add image to content":
```

**Recommended Improvement:**
- Streamlined handling of tool responses by creating dedicated methods to check tool types and manage responses.

```python
def _handle_tool_result(self, formatted_answer, tool_result):
    ...
```

## General Recommendations
1. **Type Hints**: Ensure type hints are added for all new methods to promote clarity and support for IDE features.
2. **Error Handling**: Implement more robust error handling mechanisms to ensure fault tolerance.

```python
def _execute_tool_and_check_finality(self, formatted_answer):
    ...
```

3. **Testing**: Further develop tests to cover edge cases, focusing on multimodal interactions and tool delegation scenarios.

## Historical Context
This PR marks a notable evolution in the functionality of agents within CrewAI. Past PRs have laid the groundwork for delegation strategies and modular tool usage, making these enhancements timely and necessary. Existing related PRs should be reviewed to ensure no overlapping functionalities exist, and to leverage prior learnings.

## Implications for Related Files
- **Agent Class**: The changes in the `Agent` class will require updates across any other components that interact with agents, particularly in scenarios where multimodal capabilities are integrated.
- **Tool Management Logic**: The refactoring of tool management could introduce different tool behavior if not adequately tested against legacy behavior.

## Conclusion
PR #1805 represents a meaningful improvement in the CrewAI framework by enhancing delegation and multimodal support. While the modifications are largely beneficial, the suggested recommendations focus on improving code clarity, maintainability, and system robustness, ensuring a solid foundation for future enhancements. Implementing the proposed changes will not only improve code quality but will also set a precedent for adopting best practices within the codebase moving forward. 

Overall, the changes will significantly increase the functional capacity of CrewAI agents, but attention must be given to the suggestions regarding documentation, type safety, and testing to secure long-term maintainability and effectiveness.

File: src/crewai/agents/crew_agent_executor.py [hunk 1]
```diff
@@ -143,10 +143,20 @@ def _invoke_loop(self, formatted_answer=None):
...
                             formatted_answer
                         )
-                        if self.step_callback:
-                            self.step_callback(tool_result)
 
-                        formatted_answer.text += f"\nObservation: {tool_result.result}"
+                        # Directly append the result to the messages if the
+                        # tool is "Add image to content" in case of multimodal
+                        # agents
+                        if formatted_answer.tool == "Add image to content":
```
Comment (gvieira) [inline src/crewai/agents/crew_agent_executor.py]: What about translations?

File: src/crewai/crew.py [hunk 2]
```diff
@@ -726,7 +724,14 @@ def _execute_tasks(
                     f"No agent available for task: {task.description}. Ensure that either the task has an assigned agent or a manager agent is provided."
                 )
 
-            self._prepare_agent_tools(task)
+            # Determine which tools to use - task tools take precedence over agent tools
+            tools_for_task = task.tools if task.tools else agent_to_use.tools or []
```
Comment (gvieira) [inline src/crewai/crew.py]: I'm actually surprised this `or []` works. If it does, could we safely do something like `task.tools or agent_to_use.tools or []`?

File: src/crewai/crew.py [hunk 3]
```diff
@@ -792,60 +797,82 @@ def _handle_conditional_task(
...
         if self.manager_agent:
             if task.agent:
-                self.manager_agent.tools = task.agent.get_delegation_tools([task.agent])
+                tools = self._inject_delegation_tools(tools, task.agent, [task.agent])
             else:
-                self.manager_agent.tools = self.manager_agent.get_delegation_tools(
-                    self.agents
-                )
+                tools = self._inject_delegation_tools(tools, self.manager_agent, self.agents)
+        # self.manager_agent.tools = tools
```
Comment (gvieira) [inline src/crewai/crew.py]: Not needed anymore?

File: src/crewai/tools/agent_tools/add_image_tool.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        action: str = None,
+        **kwargs,
+    ) -> dict:
+        action = action or "Please provide a detailed description of this image, including all visual elements, context, and any notable details you can observe."
+        content = [
+            {"type": "text", "text": action},
+            {
+                "type": "image_url",
+                "image_url": {
+                    "url": image_url,
```
Comment (gvieira) [inline src/crewai/tools/agent_tools/add_image_tool.py:37]: How will this work for local images? Asking cause the OpenAI Vision has a base64 strategy to handle such cases.

File: src/crewai/tools/tool_usage.py [hunk 5]
```diff
@@ -103,6 +104,19 @@ def use(
             if self.agent.verbose:
                 self._printer.print(content=f"\n\n{error}\n", color="red")
             return error
+
+        if isinstance(tool, CrewStructuredTool) and tool.name == 'Add image to content':
```
Comment (gvieira) [inline src/crewai/tools/tool_usage.py]: Still worried about translations, but actually more with having logic tied to tool name. I mean, what if someone sets up a tool with this very name?

Comment (gvieira) [reply src/crewai/tools/tool_usage.py]: Missed the translation here?

File: src/crewai/tools/agent_tools/add_image_tool.py (hunk 4 above)
Comment (joaomdmoura) [reply src/crewai/tools/agent_tools/add_image_tool.py:37]: depends on the model, for most you need to have uploaded image or base64, but that is out of our control

**Pull request merged**